src/JustSaying/Fluent/PublishConfig/DynamicAddressMessagePublisher.cs

```diff
@@ -0,0 +1,118 @@
+using System.Collections.Concurrent;
+using JustSaying.Messaging;
+using JustSaying.Messaging.Interrogation;
+using JustSaying.Models;
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying.Fluent;
+
+internal sealed class DynamicAddressMessagePublisher(
+    string topicArnTemplate,
+    Func<string, Message, string> topicAddressCustomizer,
+    Func<string, StaticAddressPublicationConfiguration> staticConfigBuilder,
+    ILoggerFactory loggerFactory) : IMessagePublisher, IMessageBatchPublisher
+{
+    private readonly string _topicArnTemplate = topicArnTemplate;
+    private readonly ConcurrentDictionary<string, IMessagePublisher> _publisherCache = new();
```

Comment:
Can we make this `ConcurrentDictionary<string, Lazy<IMessagePublisher>>`, and do away with the lock collection (`_topicCreationLocks`)? It feels simpler and safer.

Reply:
Yeah, I think that would express what we're trying to do in a nicer way, and make it less error prone to make changes to.

Reply:
In fact, we could go one further and have a `private readonly ConcurrentDictionary<string, Lazy<StaticAddressPublicationConfiguration>> _publisherCache = new();`, then we won't have to have a duplicate Dictionary for IMessagePublisher and IMessageBatchPublisher.

Reply:
Yeah, I think that would be an improvement

Reply:
I thought it was worth looking at the changes to do this in isolation, so I've raised a PR in my fork into this branch:
https://github.com/hwoodiwiss/JustSaying/pull/1

It has a side-effect of changing the result of Interrogation, but I think that's okay? It just represents the fact that we're using the same publisher for Single vs Batch publishing now